cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/datastore/DocumentStore.java

```diff
@@ -0,0 +1,192 @@
+     *                 be encrypted.
+     * @return An existing or newly created store.
+     * @throws DatastoreNotCreatedException if the database cannot be opened
+     */
+    public static DocumentStore getInstance(File location, KeyProvider provider) throws DatastoreNotCreatedException {
+        try {
+            synchronized (documentStores) {
+                DocumentStore ds = documentStores.get(location);
+                boolean created = false;
+                if (ds == null) {
+                    created = checkPathAndCreateIfNeeded(location);
+                    ds = new DocumentStore(location, provider);
+                    documentStores.put(location, ds);
+                    eventBus.post(new DatabaseOpened(ds.databaseName));
+                }
+                if (created) {
+                    eventBus.post(new DatabaseCreated(ds.databaseName));
+                }
+                return ds;
+            }
+        }
+        catch (IOException e) {
+            throw new DatastoreNotCreatedException("Database not found: " + location, e);
+        } catch (SQLException e) {
+            throw new DatastoreNotCreatedException("Database not initialized correctly: " + location, e);
+        } catch (DatastoreException e) {
+            throw new DatastoreNotCreatedException("Datastore not initialized correctly: " + location, e);
+        }
+    }
+
+    private static boolean checkPathAndCreateIfNeeded(File location) {
+        logger.fine("Datastore path: " + location);
+
+        boolean created = false;
+        if(!location.exists()){
+            created = location.mkdir();
+        }
+
+        Misc.checkArgument(location.isDirectory(), "Input path is not a valid directory");
+        Misc.checkArgument(location.canWrite(), "Datastore directory is not writable");
+        return created;
+    }
+
+    @SuppressWarnings("unchecked")
+    public void close() {
+        synchronized (documentStores) {
+            DocumentStore ds = documentStores.remove(location);
+            if (ds != null) {
+                ((DatabaseImpl)database).close();
+                ((IndexManagerImpl)query).close();
+            }
+            else {
+                throw new IllegalStateException("DocumentStore "+location+" already closed");
+            }
+        }
+        eventBus.post(new DatabaseClosed(databaseName));
+    }
+
+    public void delete() throws IOException {
+        try {
+            this.close();
+        } catch (Exception e) {
+            // caught exception could be something benign like datastore already closed, so just log
+            logger.log(Level.WARNING, "Caught exception whilst closing DocumentStore in delete()", e);
+        }
+        if (!location.exists()) {
+            String msg = String.format("Datastore %s doesn't exist on disk", location);
+            logger.warning(msg);
+            throw new IOException(msg);
+        } else {
+            FileUtils.deleteDirectory(location);
```

Comment:
If the checked exception type changes then we'll need to catch the `IOException` from a failed deletion and throw the new type here as well.